src/associations/belongs-to-many.ts

```diff
@@ -0,0 +1,1316 @@
+   *
+   * @param sourceInstance
+   * @param currentThroughRows
+   * @param newTargets
+   * @param options
+   * @private
+   */
+  async #updateAssociations(
+    sourceInstance: SourceModel,
+    currentThroughRows: ThroughModel[],
+    newTargets: TargetModel[],
+    options?:
+      & { through?: JoinTableAttributes }
+      & BulkCreateOptions<Attributes<ThroughModel>>
+      & Omit<UpdateOptions<Attributes<ThroughModel>>, 'where'>,
+  ) {
+    const sourceKey = this.sourceKey;
+    const targetKey = this.targetKey;
+    const foreignKey = this.foreignKey;
+    const otherKey = this.otherKey;
+
+    const defaultAttributes = options?.through || {};
+
+    const promises: Array<Promise<any>> = [];
+    const unassociatedTargets: TargetModel[] = [];
+    // the 'through' table of these targets has changed
+    const changedTargets: TargetModel[] = [];
+    for (const newInstance of newTargets) {
+      const existingThroughRow = currentThroughRows.find(throughRow => {
+        // @ts-expect-error -- throughRow[] instead of .get because throughRows are loaded using 'raw'
+        return throughRow[otherKey] === newInstance.get(targetKey);
+      });
+
+      if (!existingThroughRow) {
+        unassociatedTargets.push(newInstance);
+
+        continue;
+      }
+
+      // @ts-expect-error -- gets the content of the "through" table for this association that is set on the model
+      const throughAttributes = newInstance[this.through.model.name];
+      const attributes = { ...defaultAttributes, ...throughAttributes };
+
+      if (Object.keys(attributes).some(attribute => {
+        // @ts-expect-error existingThroughRow is raw
+        return attributes[attribute] !== existingThroughRow[attribute];
+      })) {
+        changedTargets.push(newInstance);
+      }
+    }
+
+    if (unassociatedTargets.length > 0) {
+      const bulk = unassociatedTargets.map(unassociatedTarget => {
+        // @ts-expect-error -- gets the content of the "through" table for this association that is set on the model
+        const throughAttributes = unassociatedTarget[this.through.model.name];
+        const attributes = { ...defaultAttributes, ...throughAttributes };
+
+        attributes[foreignKey] = sourceInstance.get(sourceKey);
+        attributes[otherKey] = unassociatedTarget.get(targetKey);
+
+        Object.assign(attributes, this.through.scope);
+
+        return attributes;
+      });
+
+      promises.push(this.through.model.bulkCreate(bulk, { validate: true, ...options }));
+    }
+
+    for (const changedTarget of changedTargets) {
+      // @ts-expect-error -- gets the content of the "through" table for this association that is set on the model
+      let throughAttributes = changedTarget[this.through.model.name];
+      const attributes = { ...defaultAttributes, ...throughAttributes };
+      // Quick-fix for subtle bug when using existing objects that might have the through model attached (not as an attribute object)
```

Comment:
Do we want to make this a TODO for a future PR if somebody wants to fix this properly?